# Biology {#biology}

## General
Some helpful commands for genetic sequence.
```{python eval=FALSE}
from string import ascii_uppercase # python 3
from string import upper, lower # python 2
upper('tcga')
lower('TCGA')
title('tcga') # capitalize the first letter
```

## Biopython 
Reverse complement of sequence
```{python eval=FALSE}
from Bio.Seq import Seq
str(Seq(i).reverse_complement())
```

## UCSC Genome Browser
Get sequence from UCSC genome browser
```{python eval=FALSE}
from subprocess import check_output, STDOUT
temp = check_output('wget -qO- http://genome.ucsc.edu/cgi-bin/das/hg19/dna?segment=%s:%s,%s' % (vcfObj.chrom,low,high), stderr=STDOUT, shell=True)
```

Annotate genomic locations using the UCSC genome browser can be done in bash by getting the refGene track information.
```{python eval=FALSE}
wget -qO- "https://api.genome.ucsc.edu/getData/track?genome=hg38;track=refGene;chrom=chr8;start=127736600;end=127736601"
```

The following code just returns the gene name for an associated region in python.
```{python eval=FALSE}
from subprocess import check_output, STDOUT
import json
chrom = '8'
start = '127736600'
stop = '127736601'
api = 'https://api.genome.ucsc.edu/getData/track?genome=hg38;track=refGene;'

temp = json.loads(check_output(
    'wget -qO- "{api}chrom=chr{chrom};start={start};end={stop}"'.format(
        api=api, chrom=chrom, start=start, stop=stop),
        stderr=STDOUT, shell=True))['refGene'][0]['name2']
```

And here is a python implementation to annotate a pandas DataFrame.
```{python eval=FALSE}
def annotate(df):

    from subprocess import check_output, STDOUT
    import json

    genes = []
    for index, row in df.iterrows():

        chrom = row.Chrom
        start = row.Loc
        stop = row.Loc + 1
        api = 'https://api.genome.ucsc.edu/getData/track?genome=hg38;track=refGene;'

        gene = json.loads(check_output(
            'wget -qO- "{api}chrom=chr{chrom};start={start};end={stop}"'.format(
                api=api, chrom=chrom, start=start, stop=stop),
                stderr=STDOUT, shell=True))['refGene']
        if not len(gene) == 0: gene = gene[0]['name2']
        else: gene = np.nan

        genes.append(gene)

    df['Gene'] = genes


    return df
```

If the reference genome comes in the .2bit format, it is likely that it should then be converted to .fa format, and .2bittofa can accomplish this.
```{python eval=FALSE}
twoBitToFa refgenomes/hg19.2bit refgenomes/hg19.fa
```

The newly created .fa file will need to be indexed if it is to be used with gatk and bwa mem.

Here is an example of indexing where -a bwtsw specifies that we want to use the indexing algorithm that is capable of handling the whole human genome.
```{python eval=FALSE}
bwa index -a bwtsw reference.fa
```

Then create a file called reference.fa.fai, with one record per line for each of the contigs in the FASTA reference file. Each record is composed of the contig name, size, location, basesPerLine and bytesPerLine.
```{python eval=FALSE}
samtools faidx reference.fa
```

## Ref Genome
Get sequence from reference genome
```{python eval=FALSE}
from subprocess import check_output, STDOUT
temp = check_output('samtools faidx %s %s:%s-%s' % (ref, vcfObj.chrom, low, high), stderr=STDOUT, shell=True)

finalSeq = ''
for line in temp.decode('UTF-8').split('\n'):
for line in temp.decode('UTF-8').split('\n'): # this is only necessary in python 3 to convert binary to string
	if '>' not in line:
		finalSeq += line

finalSeq = finalSeq.upper()
```

## Personal Information
```{python eval=FALSE}
# parse vcf file with parseline
if '#' not in line and 'chr' in line: # skip the info
# vcf handling
from parseline import VCFObj
# or
from util import VCFObj
vcfObj = VCFObj(vcfLine)
# available attributes: ao, dp, af, wt, var, chrom, location
```

## GATK
The entire set of GATK software can be downloaded together [here]('https://software.broadinstitute.org/gatk/').  
Methods and algorithms as they are run at the Broad can be found [here]('https://software.broadinstitute.org/gatk/documentation/topic?name=methods').  
The Broad also has a number of [tutorials]('https://software.broadinstitute.org/gatk/documentation/topic?name=tutorials') that introduce some of the software packaged within GATK. 

### Mutect2
A [tutorial]('https://software.broadinstitute.org/gatk/documentation/article?id=24057') exists that introduces how Mutect2 can be used to call somatic mutations.  
In general Mutect2 can be run as in the following.  
```{python eval=FALSE}
../util/gatk-4.1.4.0/gatk Mutect2 -R /seq/references/Homo_sapiens_assembly38/v0/Homo_sapiens_assembly38.fasta -I tumor.bam -O unfiltered.vcf
```

## Samtools                                         
Grab a single chromosomal region from a `.bam` file.
```{python eval=FALSE}                              
samtools view -b in.bam chr1 > in_chr1.bam          
```                                                 


