# Bioinformatics Resources {#resources}

## Cancer Datasets
* [CancerMine](http://bionlp.bcgsc.ca/cancermine/) is a site that mines publication data to create a database of mutations labeled as drivers, oncogenes, or tumor suppressors. These classifications may help to understand the evolution of different cancers.
* [The Cancer Cell Line Encyclopedia](https://www.nature.com/articles/s41586-019-1186-3?WT.ec_id=NATURE-201905&sap-outbound-id=720FBE4668F2FE14A299D42CB43EF50D90063A16&mkt-key=005056A5C6311ED999AC3A52796F7641) has a wealth of information from a large number of cancer cell lines.
* [Project score](https://score.depmap.sanger.ac.uk/) has a number of genetic screens that may be useful in identifying pathways that are critical to cancer growth and survival
* [cbioportal](http://www.cbioportal.org/) has information about codon changes in cancer but does not seem to have any sequence data
* [dbGAP](https://dbgap.ncbi.nlm.nih.gov/aa/wga.cgi?page=login) has info on genotypes and phenotypes, whatever the fuck that means
	+ [Here](https://dbgap.ncbi.nlm.nih.gov/aa/wga.cgi?page=login) is the page with instructions on how to get dbGAP access
* [TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga)
	+ [This](https://gdc.cancer.gov/access-data/obtaining-access-controlled-data) page has the instructions on how to get TCGA access
* [COSMIC](https://cancer.sanger.ac.uk/cosmic)
	+ Cosmic has a project called the [Cancer Gene Census](https://cancer.sanger.ac.uk/cosmic/census?genome=37) in which they are trying to catalog all mutations that have been implicated in playing a causal role in cancer
	+ They also have implemented convenient [ways](https://cancer.sanger.ac.uk/cosmic/help/file_download) of directly downloading information and files from the database in python using the files hosted [here](https://cancer.sanger.ac.uk/cosmic/download).
* [Mastermind](https://mastermind.genomenon.com/api) from genomenon parsed all the articles on pubmed in order to find any and all information for each possible mutation
* [GTEx](https://gtexportal.org/home/) has RNA-Seq, Exome Seq, WGseq, SNP arrays, gene expression arrays and more for cancer and non-cancer?
* [TARGET](https://ocg.cancer.gov/programs/target) is a dataset at the NIH that has childhood cancer data.

## Alzheimer's
* [The Alzheimers sequencing project](https://www.niagads.org/adsp/content/home) is gathering data to understand late onset alzheimer's

## Genome Resources
* Jax has a human to mouse gene matching [list](http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt) that provides gene location matching between human and mouse
* UCSC genome browser has an [API](http://genomewiki.ucsc.edu/index.php/Programmatic_access_to_the_Genome_Browser) for programmatic access

## Genomic Datasets 
* [AllOfUs](ttps://www.joinallofus.org/en/how-to-join) is sequencing and collecting other health data on a million individuals
* [Color Genomics](https://www.color.com/) is one of three companies that will be doing the sequencing and testing for AllOfUs
* [1000 genomes](http://www.internationalgenome.org/data/)
* [100,000 genomes](https://www.genomicsengland.co.uk/)

## Computing Tools
* [AWS](https://aws.amazon.com/health/) looks like it has some healthcare and life sciences resources

## Microbiome Datasets
* [The Human Microbiome Project Data Portal](https://portal.hmpdacc.org/) from Michael Snyder's group has longitudinal 'omics data that includes diseased and healthy timepoints.
* Michael Snyder's [iPOP Personal 'Omics Profiling](http://med.stanford.edu/ipop.html) has some interesting microbiome data specifically targeted to understanding diabetes.
* People respond differently to different drugs, and this appears to in part be due to the differential drug metabolism of their gut microbiome. Some of the differences can be observed when different strains of gut bacteria are isolated and directly exposed to drug to understand how the drugs are differentially metabolized [@zimmermann2019mapping]. The bacterial sequencing data is available [here](https://www.ebi.ac.uk/ena/data/search?query=PRJEB31790), and some of the extra drug
  screening data is available [here](https://figshare.com/articles/Mapping_human_microbiome_drug_metabolism_by_gut_bacteria_and_their_genes/8119058).
* Multi-omics of Crohn's and ulcerative colitis can be found as part of the [Integrative Human Microbiome Project](https://www.ibdmdb.org/).

## eQTL/RNASeq and other Tools
* [Here](https://adinasarapu.github.io/year-archive/) are a good number of general tools to check out
* [Here](https://github.com/molgenis/systemsgenetics/wiki/eQTL-mapping-analysis-cookbook-for-RNA-seq-data) is a good python eQTL analysis
* [genenetwork2](http://gn2.genenetwork.org/) has a great deal of data and eQTL mapping [tools](https://github.com/genenetwork/genenetwork2)
* The Eske Derks lab has some [hosted data](https://github.com/MareesAT/GWA_tutorial/) that walks through how to properly run a GWAS analysis [@Marees2018-rc].

## Cell Biology
* [HuBMAP](https://hubmapconsortium.org/) is a group that is organizing tissues at the single-cell scale to map tissues and organisms spatially.

## Other Data Resources
* [The Earth Microbiome Project](http://www.earthmicrobiome.org/data-and-code/) already has data available, and is trying to sequence all non-eukaryotic life on earth
    + They have detailed information about the project in their [PNAS paper](http://www.pnas.org/content/115/17/4325.long) from 2018
* [Information is Beautiful](https://informationisbeautiful.net/) has some interesting datasets
