<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data Science With Python</title>
  <meta name="description" content="These are notes and thoughts within the scope of the field of genomics.">
  <meta name="generator" content="bookdown <!--bookdown:version--> and GitBook 2.6.7">

  <meta property="og:title" content="Data Science With Python" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="These are notes and thoughts within the scope of the field of genomics." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data Science With Python" />
  
  <meta name="twitter:description" content="These are notes and thoughts within the scope of the field of genomics." />
  

<meta name="author" content="L A Liggett">


<meta name="date" content="2019-06-08">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">Data Science With Python</h1>
<p class="author"><em>L A Liggett</em></p>
<p class="date"><em>2019-06-08</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#prerequisites"><span class="toc-section-number">1</span> Prerequisites</a></li>
<li><a href="#intro"><span class="toc-section-number">2</span> Introduction</a></li>
<li><a href="#jupyter"><span class="toc-section-number">3</span> JupyterLab</a></li>
<li><a href="#visualization"><span class="toc-section-number">4</span> Visualization</a><ul>
<li><a href="#color"><span class="toc-section-number">4.1</span> Color</a><ul>
<li><a href="#colorschemes"><span class="toc-section-number">4.1.1</span> Colorschemes</a></li>
<li><a href="#controlling-coloration"><span class="toc-section-number">4.1.2</span> Controlling Coloration</a></li>
</ul></li>
<li><a href="#matplotlib"><span class="toc-section-number">4.2</span> Matplotlib</a></li>
<li><a href="#seaborn"><span class="toc-section-number">4.3</span> Seaborn</a></li>
<li><a href="#statistics"><span class="toc-section-number">4.4</span> Statistics</a></li>
<li><a href="#various-plot-styles"><span class="toc-section-number">4.5</span> Various Plot Styles</a></li>
</ul></li>
<li><a href="#biology"><span class="toc-section-number">5</span> Biology</a><ul>
<li><a href="#general"><span class="toc-section-number">5.1</span> General</a></li>
<li><a href="#biopython"><span class="toc-section-number">5.2</span> Biopython</a></li>
<li><a href="#ucsc-genome-browser"><span class="toc-section-number">5.3</span> UCSC Genome Browser</a></li>
<li><a href="#ref-genome"><span class="toc-section-number">5.4</span> Ref Genome</a></li>
<li><a href="#personal-information"><span class="toc-section-number">5.5</span> Personal Information</a></li>
</ul></li>
<li><a href="#io"><span class="toc-section-number">6</span> Data I/O</a><ul>
<li><a href="#reading-data-files"><span class="toc-section-number">6.1</span> Reading Data Files</a></li>
<li><a href="#pickles"><span class="toc-section-number">6.2</span> Pickles</a></li>
</ul></li>
<li><a href="#pandas"><span class="toc-section-number">7</span> Pandas</a><ul>
<li><a href="#file-io"><span class="toc-section-number">7.1</span> File I/O</a></li>
<li><a href="#relabeling"><span class="toc-section-number">7.2</span> Relabeling</a></li>
<li><a href="#sorting-and-arranging"><span class="toc-section-number">7.3</span> Sorting and Arranging</a></li>
<li><a href="#editing-data"><span class="toc-section-number">7.4</span> Editing Data</a><ul>
<li><a href="#replace-values"><span class="toc-section-number">7.4.1</span> Replace values</a></li>
</ul></li>
<li><a href="#combining-data-structures"><span class="toc-section-number">7.5</span> Combining Data Structures</a></li>
<li><a href="#splitting"><span class="toc-section-number">7.6</span> Splitting</a></li>
<li><a href="#summarizing"><span class="toc-section-number">7.7</span> Summarizing</a></li>
<li><a href="#arithmetic-and-row-wise-analysis"><span class="toc-section-number">7.8</span> Arithmetic and Row-wise Analysis</a></li>
</ul></li>
<li><a href="#git"><span class="toc-section-number">8</span> Git</a><ul>
<li><a href="#setup"><span class="toc-section-number">8.1</span> Setup</a><ul>
<li><a href="#git-setup"><span class="toc-section-number">8.1.1</span> Git Setup</a></li>
<li><a href="#repository-initiation"><span class="toc-section-number">8.1.2</span> Repository Initiation</a></li>
<li><a href="#mirror-on-online-repository"><span class="toc-section-number">8.1.3</span> Mirror on Online Repository</a></li>
</ul></li>
<li><a href="#manipulating-commits"><span class="toc-section-number">8.2</span> Manipulating Commits</a><ul>
<li><a href="#repository-status"><span class="toc-section-number">8.2.1</span> Repository Status</a></li>
<li><a href="#file-checkout"><span class="toc-section-number">8.2.2</span> File Checkout</a></li>
<li><a href="#resetting-a-repository"><span class="toc-section-number">8.2.3</span> Resetting a Repository</a></li>
</ul></li>
</ul></li>
<li><a href="#git"><span class="toc-section-number">9</span> Git</a><ul>
<li><a href="#formatting"><span class="toc-section-number">9.1</span> Formatting</a></li>
<li><a href="#spellcheck"><span class="toc-section-number">9.2</span> Spellcheck</a><ul>
<li><a href="#git-setup-1"><span class="toc-section-number">9.2.1</span> Git Setup</a></li>
<li><a href="#repository-initiation-1"><span class="toc-section-number">9.2.2</span> Repository Initiation</a></li>
<li><a href="#mirror-on-online-repository-1"><span class="toc-section-number">9.2.3</span> Mirror on Online Repository</a></li>
</ul></li>
<li><a href="#manipulating-commits-1"><span class="toc-section-number">9.3</span> Manipulating Commits</a><ul>
<li><a href="#repository-status-1"><span class="toc-section-number">9.3.1</span> Repository Status</a></li>
<li><a href="#file-checkout-1"><span class="toc-section-number">9.3.2</span> File Checkout</a></li>
<li><a href="#resetting-a-repository-1"><span class="toc-section-number">9.3.3</span> Resetting a Repository</a></li>
</ul></li>
</ul></li>
<li><a href="#references">References</a></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data Science With Python</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="prerequisites" class="section level1">
<h1><span class="header-section-number">1</span> Prerequisites</h1>
<p>This is a <em>sample</em> book written in <strong>Markdown</strong>. You can use anything that Pandocâ€™s Markdown supports, e.g., a math equation <span class="math inline">\(a^2 + b^2 = c^2\)</span>.</p>
<p>The <strong>bookdown</strong> package can be installed from CRAN or Github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;bookdown&quot;</span>)
<span class="co"># or the development version</span>
<span class="co"># devtools::install_github(&quot;rstudio/bookdown&quot;)</span></code></pre></div>
<p>Remember each Rmd file contains one and only one chapter, and a chapter is defined by the first-level heading <code>#</code>.</p>
<p>To compile this example to PDF, you need XeLaTeX. You are recommended to install TinyTeX (which includes XeLaTeX): <a href="https://yihui.name/tinytex/" class="uri">https://yihui.name/tinytex/</a>.</p>
<!--chapter:end:index.Rmd-->
</div>
<div id="intro" class="section level1">
<h1><span class="header-section-number">2</span> Introduction</h1>
<p>You can label chapter and section titles using <code>{#label}</code> after them, e.g., we can reference Chapter @ref(intro). If you do not manually label them, there will be automatic labels anyway, e.g., Chapter @ref(visualization).</p>
<p>Figures and tables with captions will be placed in <code>figure</code> and <code>table</code> environments, respectively.</p>
<p>And this is some other random stuff.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">4</span>, .<span class="dv">1</span>, .<span class="dv">1</span>))
<span class="kw">plot</span>(pressure, <span class="dt">type =</span> <span class="st">&#39;b&#39;</span>, <span class="dt">pch =</span> <span class="dv">19</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="bookdown-demo_files/figure-html/nice-fig-1.png" alt="Here is a nice figure!" width="80%" />
<p class="caption">
(#fig:nice-fig)Here is a nice figure!
</p>
</div>
<p>Reference a figure by its code chunk label with the <code>fig:</code> prefix, e.g., see Figure @ref(fig:nice-fig). Similarly, you can reference tables generated from <code>knitr::kable()</code>, e.g., see Table @ref(tab:nice-tab).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw">kable</span>(
  <span class="kw">head</span>(iris, <span class="dv">20</span>), <span class="dt">caption =</span> <span class="st">&#39;Here is a nice table!&#39;</span>,
  <span class="dt">booktabs =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<table>
<caption>(#tab:nice-tab)Here is a nice table!</caption>
<thead>
<tr class="header">
<th align="right">Sepal.Length</th>
<th align="right">Sepal.Width</th>
<th align="right">Petal.Length</th>
<th align="right">Petal.Width</th>
<th align="left">Species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">5.1</td>
<td align="right">3.5</td>
<td align="right">1.4</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">4.9</td>
<td align="right">3.0</td>
<td align="right">1.4</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">4.7</td>
<td align="right">3.2</td>
<td align="right">1.3</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">4.6</td>
<td align="right">3.1</td>
<td align="right">1.5</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">5.0</td>
<td align="right">3.6</td>
<td align="right">1.4</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">5.4</td>
<td align="right">3.9</td>
<td align="right">1.7</td>
<td align="right">0.4</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">4.6</td>
<td align="right">3.4</td>
<td align="right">1.4</td>
<td align="right">0.3</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">5.0</td>
<td align="right">3.4</td>
<td align="right">1.5</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">4.4</td>
<td align="right">2.9</td>
<td align="right">1.4</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">4.9</td>
<td align="right">3.1</td>
<td align="right">1.5</td>
<td align="right">0.1</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">5.4</td>
<td align="right">3.7</td>
<td align="right">1.5</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">4.8</td>
<td align="right">3.4</td>
<td align="right">1.6</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">4.8</td>
<td align="right">3.0</td>
<td align="right">1.4</td>
<td align="right">0.1</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">4.3</td>
<td align="right">3.0</td>
<td align="right">1.1</td>
<td align="right">0.1</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">5.8</td>
<td align="right">4.0</td>
<td align="right">1.2</td>
<td align="right">0.2</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">5.7</td>
<td align="right">4.4</td>
<td align="right">1.5</td>
<td align="right">0.4</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">5.4</td>
<td align="right">3.9</td>
<td align="right">1.3</td>
<td align="right">0.4</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">5.1</td>
<td align="right">3.5</td>
<td align="right">1.4</td>
<td align="right">0.3</td>
<td align="left">setosa</td>
</tr>
<tr class="odd">
<td align="right">5.7</td>
<td align="right">3.8</td>
<td align="right">1.7</td>
<td align="right">0.3</td>
<td align="left">setosa</td>
</tr>
<tr class="even">
<td align="right">5.1</td>
<td align="right">3.8</td>
<td align="right">1.5</td>
<td align="right">0.3</td>
<td align="left">setosa</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw">include_graphics</span>(<span class="kw">rep</span>(<span class="st">&quot;knit-logo.png&quot;</span>, <span class="dv">3</span>))</code></pre></div>
<p><img src="knit-logo.png" /><!-- --><img src="knit-logo.png" /><!-- --><img src="knit-logo.png" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw">include_app</span>(<span class="st">&quot;https://yihui.shinyapps.io/miniUI/&quot;</span>, 
                     <span class="dt">height =</span> <span class="st">&quot;600px&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> pandas <span class="im">as</span> pd
x <span class="op">=</span> <span class="st">&#39;hello, python world!&#39;</span>
<span class="bu">print</span>(x.split(<span class="st">&#39; &#39;</span>))</code></pre></div>
<pre><code>## [&#39;hello,&#39;, &#39;python&#39;, &#39;world!&#39;]</code></pre>
<p>You can write citations, too. For example, we are using the <strong>bookdown</strong> package <span class="citation">(Xie <a href="#ref-R-bookdown">2018</a>)</span> in this sample book, which was built on top of R Markdown and <strong>knitr</strong> <span class="citation">(<span class="citeproc-not-found" data-reference-id="xie2015"><strong>???</strong></span>)</span>.</p>
<!--chapter:end:01-intro.Rmd-->
</div>
<div id="jupyter" class="section level1">
<h1><span class="header-section-number">3</span> JupyterLab</h1>
<p>Here is a simple template that I use that controls a couple useful things when starting a new notebook.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> sys
sys.path.append(<span class="st">&#39;../util&#39;</span>)

<span class="op">%</span>reload_ext autoreload
<span class="op">%</span>autoreload <span class="dv">2</span>

<span class="im">from</span> util <span class="im">import</span> <span class="op">*</span>
<span class="im">import</span> numpy <span class="im">as</span> np                  
<span class="im">import</span> pandas <span class="im">as</span> pd                 
<span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt
<span class="im">import</span> seaborn <span class="im">as</span> sns

sns.set_palette(<span class="st">&#39;pastel&#39;</span>)
sns.set_style(<span class="st">&#39;ticks&#39;</span>)
sns.set_context(<span class="st">&#39;paper&#39;</span>, font_scale<span class="op">=</span><span class="dv">1</span>)</code></pre></div>
<p>It is often convenient to have a notebook automatically refresh the imported libraries so that they can be modified while working on a JupyterLab notebook.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="op">%</span>reload_ext autoreload
<span class="op">%</span>autoreload <span class="dv">2</span></code></pre></div>
<p>To allow directory organization, dependcies can be separated into different directories and imported into a jupyter notebook using the following import statement.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> sys
sys.path.append(<span class="st">&#39;../util&#39;</span>)</code></pre></div>
<p>A table of contents can be created to refer to each of the headers throughout a notebook in html format. The code is below (Obviously needs to be simplified.)</p>
<div class="sourceCode"><pre class="sourceCode html"><code class="sourceCode html"><span class="kw">&lt;h1&gt;</span>Table of Contents<span class="kw">&lt;span</span><span class="ot"> class=</span><span class="st">&quot;tocSkip&quot;</span><span class="kw">&gt;&lt;/span&gt;&lt;/h1&gt;</span>
<span class="kw">&lt;div</span><span class="ot"> class=</span><span class="st">&quot;toc&quot;</span><span class="kw">&gt;</span>
    <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
    <span class="kw">&lt;li&gt;</span>
        <span class="kw">&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Python-Setup&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Python-Setup-1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Python Setup<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
        <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
    <span class="kw">&lt;li&gt;</span>
        <span class="kw">&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Change-the-width-of-the-page&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Change-the-width-of-the-page-1.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>1.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Change the width of the page<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;</span>
            <span class="kw">&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Import-packages&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Import-packages-1.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>1.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Import packages<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;/ul&gt;</span>
    <span class="kw">&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;</span>
        <span class="kw">&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Colours&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Colours-2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Colours<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
        <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Colour-line-graph&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Colour-line-graph-2.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>2.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Colour line graph<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;/ul&gt;</span>
    <span class="kw">&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Totals-for-studies&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Totals-for-studies-3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Totals for studies<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Functions-for-calculating-trinucleotide-context-specific-mutation-rates&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Functions-for-calculating-trinucleotide-context-specific-mutation-rates-4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Functions for calculating trinucleotide-context specific mutation rates<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
        <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Calculating-mutation-rates-for-individual-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Calculating-mutation-rates-for-individual-variants-4.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Calculating mutation rates for individual variants<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
            <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#DNMT3A&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;DNMT3A-4.1.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.1.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>DNMT3A<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TET2&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TET2-4.1.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.1.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TET2<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#ASXL1&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;ASXL1-4.1.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.1.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>ASXL1<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TP53&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TP53-4.1.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.1.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TP53<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Calculating-mutation-rates-from-a-.csv-file-of-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Calculating-mutation-rates-from-a-.csv-file-of-variants-4.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Calculating mutation rates from a .csv file of variants<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
            <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#DNMT3A&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;DNMT3A-4.2.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.2.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>DNMT3A<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TET2&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TET2-4.2.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.2.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TET2<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#ASXL1&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;ASXL1-4.2.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.2.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>ASXL1<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TP53&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TP53-4.2.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.2.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TP53<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Calculating-mutation-rates-from-a-list-of-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Calculating-mutation-rates-from-a-list-of-variants-4.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Calculating mutation rates from a list of variants<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
            <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#DNMT3A&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;DNMT3A-4.3.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.3.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>DNMT3A<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TET2&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TET2-4.3.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.3.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TET2<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#ASXL1&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;ASXL1-4.3.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.3.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>ASXL1<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TP53&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TP53-4.3.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>4.3.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TP53<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Lists-of-variants-targeted-by-each-study&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Lists-of-variants-targeted-by-each-study-5&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Lists of variants targeted by each study<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Jaiswal-2014&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Jaiswal-2014-5.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Jaiswal 2014<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Genovese-2014&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Genovese-2014-5.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Genovese 2014<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#McKerrel-2015&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;McKerrel-2015-5.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>McKerrel 2015<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Zink-2017&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Zink-2017-5.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Zink 2017<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Coombs-2017&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Coombs-2017-5.5&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.5<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Coombs 2017<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Young-2016-</span><span class="dv">&amp;amp;</span><span class="st">-2019&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Young-2016-</span><span class="dv">&amp;amp;</span><span class="st">-2019-5.6&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.6<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Young 2016 <span class="dv">&amp;amp;</span> 2019<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Desai-2018&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Desai-2018-5.7&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.7<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Desai 2018<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Acuna-Hidalgo-2017&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Acuna-Hidalgo-2017-5.8&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>5.8<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Acuna-Hidalgo 2017<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53-6&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>6<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
        <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#DNMT3A&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;DNMT3A-6.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>6.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>DNMT3A<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TET2&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TET2-6.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>6.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TET2<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#ASXL1&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;ASXL1-6.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>6.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>ASXL1<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TP53&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TP53-6.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>6.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TP53<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Actual-number-of-observations-of-each-variant&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Actual-number-of-observations-of-each-variant-7&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>7<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Actual number of observations of each variant<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
        <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
        <span class="kw">&lt;li&gt;</span>
            <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#DNMT3A&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;DNMT3A-7.0.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>7.0.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>DNMT3A<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
            <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TET2&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TET2-7.0.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>7.0.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TET2<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#ASXL1&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;ASXL1-7.0.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>7.0.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>ASXL1<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TP53&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TP53-7.0.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>7.0.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TP53<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Functions-for-calculating-the-expected-number-of-observations-of-a-variant&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Functions-for-calculating-the-expected-number-of-observations-of-a-variant-8&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>8<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Functions for calculating the expected number of observations of a variant<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
    <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#Maximum-Likelihood-Estimation-for-s&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;Maximum-Likelihood-Estimation-for-s-9&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>9<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>Maximum Likelihood Estimation for s<span class="kw">&lt;/a&gt;&lt;/span&gt;</span>
        <span class="kw">&lt;ul</span><span class="ot"> class=</span><span class="st">&quot;toc-item&quot;</span><span class="kw">&gt;&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#DNMT3A-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;DNMT3A-variants-9.1&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>9.1<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>DNMT3A variants<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TET2-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TET2-variants-9.2&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>9.2<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TET2 variants<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#ASXL1-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;ASXL1-variants-9.3&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>9.3<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>ASXL1 variants<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;li&gt;&lt;span&gt;&lt;a</span><span class="ot"> href=</span><span class="st">&quot;#TP53-variants&quot;</span><span class="ot"> data-toc-modified-id=</span><span class="st">&quot;TP53-variants-9.4&quot;</span><span class="kw">&gt;&lt;span</span><span class="ot"> class=</span><span class="st">&quot;toc-item-num&quot;</span><span class="kw">&gt;</span>9.4<span class="dv">&amp;nbsp;&amp;nbsp;</span><span class="kw">&lt;/span&gt;</span>TP53 variants<span class="kw">&lt;/a&gt;&lt;/span&gt;&lt;/li&gt;</span>
        <span class="kw">&lt;/ul&gt;</span>
        <span class="kw">&lt;/li&gt;</span>
    <span class="kw">&lt;/ul&gt;</span>
<span class="kw">&lt;/div&gt;</span></code></pre></div>
<!--chapter:end:02-jupyter.Rmd-->
</div>
<div id="visualization" class="section level1">
<h1><span class="header-section-number">4</span> Visualization</h1>
<div id="color" class="section level2">
<h2><span class="header-section-number">4.1</span> Color</h2>
<div id="colorschemes" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Colorschemes</h3>
<p>Seaborn Themes</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">Pastel: {<span class="st">&#39;Blue&#39;</span>:<span class="st">&#39;#a3c6ff&#39;</span>, <span class="st">&#39;Orange&#39;</span>:<span class="st">&#39;#f7ab60&#39;</span>, <span class="st">&#39;Green&#39;</span>:<span class="st">&#39;#60f7a9&#39;</span>, <span class="st">&#39;Red&#39;</span>:<span class="st">&#39;#fc9d94&#39;</span>, <span class="st">&#39;Purple&#39;</span>:<span class="st">&#39;#bea3ff&#39;</span>, <span class="st">&#39;Brown&#39;</span>:<span class="st">&#39;#d1b485&#39;</span>, <span class="st">&#39;Pink&#39;</span>:<span class="st">&#39;#f7afdf&#39;</span>, <span class="st">&#39;Gray&#39;</span>:<span class="st">&#39;#c4c4c4&#39;</span>, <span class="st">&#39;Yellow&#39;</span>:<span class="st">&#39;#ffffaa&#39;</span>, <span class="st">&#39;LBlue&#39;</span>:<span class="st">&#39;#baf6ff&#39;</span>}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">Deep: {<span class="st">&#39;Green&#39;</span>:<span class="st">&#39;#5baf68&#39;</span>}</code></pre></div>
</div>
<div id="controlling-coloration" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Controlling Coloration</h3>
<p>Not all plots automatically plot with a white background, and when using something dark like jupyterlab or a presentation this can be frustrating. The background color can be set in pyplot like this.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">fig.patch.set_facecolor(<span class="st">&#39;xkcd:mint green&#39;</span>)</code></pre></div>
<p>When plotting, samples will not always be colored with the same color, especially when different subsets of samples are included in different plots. Here is a manual workaround to specify the coloration of displayed data. This is a bit cumbersome so there might be a more elegant way of achieving the same outcome.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="co"># here is an example where sample order is controlled from a pandas DataFrame</span>
sample_order <span class="op">=</span> all_vars.sort_values([<span class="st">&#39;ID&#39;</span>]).drop_duplicates([<span class="st">&#39;Sample&#39;</span>]).Sample

<span class="co"># the color order is specified here</span>
<span class="co"># colors should be in the same order as the above sample_order Series, excluding samples with no data</span>
colors <span class="op">=</span> [pastel[<span class="st">&#39;Brown&#39;</span>], pastel[<span class="st">&#39;Blue&#39;</span>],
          pastel[<span class="st">&#39;Orange&#39;</span>], pastel[<span class="st">&#39;Purple&#39;</span>],
          pastel[<span class="st">&#39;Green&#39;</span>], pastel[<span class="st">&#39;Red&#39;</span>],
          ]

plt.figure()
<span class="co"># this is an example of plotting that uses the sample_order and palette to control coloration order</span>
sns.catplot(x<span class="op">=</span><span class="st">&#39;Sample&#39;</span>, y<span class="op">=</span><span class="st">&#39;VAF&#39;</span>, hue<span class="op">=</span><span class="st">&#39;Gene&#39;</span>, jitter<span class="op">=</span><span class="va">True</span>,
            data<span class="op">=</span>oncogenic[oncogenic.Location <span class="op">==</span> <span class="st">&#39;Peripheral&#39;</span>],
            legend<span class="op">=</span><span class="va">False</span>, order<span class="op">=</span>sample_order, palette<span class="op">=</span>sns.color_palette(colors))

<span class="co"># a colorscheme can be specified if desired</span>
pastel <span class="op">=</span> {<span class="st">&#39;Blue&#39;</span>:<span class="st">&#39;#a3c6ff&#39;</span>, <span class="st">&#39;Orange&#39;</span>:<span class="st">&#39;#f7ab60&#39;</span>,
          <span class="st">&#39;Green&#39;</span>:<span class="st">&#39;#60f7a9&#39;</span>, <span class="st">&#39;Red&#39;</span>:<span class="st">&#39;#fc9d94&#39;</span>,
          <span class="st">&#39;Purple&#39;</span>:<span class="st">&#39;#bea3ff&#39;</span>, <span class="st">&#39;Brown&#39;</span>:<span class="st">&#39;#d1b485&#39;</span>,
          <span class="st">&#39;Pink&#39;</span>:<span class="st">&#39;#f7afdf&#39;</span>, <span class="st">&#39;Gray&#39;</span>:<span class="st">&#39;#c4c4c4&#39;</span>,
          <span class="st">&#39;Yellow&#39;</span>:<span class="st">&#39;#ffffaa&#39;</span>, <span class="st">&#39;LBlue&#39;</span>:<span class="st">&#39;#baf6ff&#39;</span>}

<span class="co"># this controls the coloration in the legend</span>
<span class="im">import</span> matplotlib.patches <span class="im">as</span> mpatches
egfr <span class="op">=</span> mpatches.Patch(color<span class="op">=</span>pastel[<span class="st">&#39;Blue&#39;</span>], label<span class="op">=</span><span class="st">&#39;EGFR&#39;</span>)
pik3ca <span class="op">=</span> mpatches.Patch(color<span class="op">=</span>pastel[<span class="st">&#39;Orange&#39;</span>], label<span class="op">=</span><span class="st">&#39;PIK3CA&#39;</span>)
myc <span class="op">=</span> mpatches.Patch(color<span class="op">=</span>pastel[<span class="st">&#39;Green&#39;</span>], label<span class="op">=</span><span class="st">&#39;MYC&#39;</span>)

plt.legend(handles<span class="op">=</span>[egfr,pik3ca,myc],
           loc<span class="op">=</span><span class="st">&#39;upper right&#39;</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">1.5</span>, <span class="dv">1</span>),
           ncol<span class="op">=</span><span class="dv">1</span>) <span class="co"># no legend overlap</span></code></pre></div>
</div>
</div>
<div id="matplotlib" class="section level2">
<h2><span class="header-section-number">4.2</span> Matplotlib</h2>
<p>Plotting a heatmap.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt
<span class="im">import</span> numpy <span class="im">as</span> np
a <span class="op">=</span> np.random.random((<span class="dv">16</span>, <span class="dv">16</span>))
plt.imshow(a, cmap<span class="op">=</span><span class="st">&#39;RdBu&#39;&#39;, interpolation=&#39;</span>nearest<span class="st">&#39;)</span>
<span class="st">plt.show()</span></code></pre></div>
<p>Possible heatmap colors are:</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">Accent, Accent_r, Blues, Blues_r, BrBG, BrBG_r, BuGn, BuGn_r, BuPu, BuPu_r, CMRmap, CMRmap_r, Dark2, Dark2_r, GnBu, GnBu_r, Greens, Greens_r, Greys, Greys_r, OrRd, OrRd_r, Oranges, Oranges_r, PRGn, PRGn_r, Paired, Paired_r, Pastel1, Pastel1_r, Pastel2, Pastel2_r, PiYG, PiYG_r, PuBu, PuBuGn, PuBuGn_r, PuBu_r, PuOr, PuOr_r, PuRd, PuRd_r, Purples, Purples_r, RdBu, RdBu_r, RdGy, RdGy_r, RdPu, RdPu_r, RdYlBu, RdYlBu_r, RdYlGn, RdYlGn_r, Reds, Reds_r, Set1,
Set1_r, Set2, Set2_r, Set3, Set3_r, Spectral, Spectral_r, Wistia, Wistia_r, YlGn, YlGnBu, YlGnBu_r, YlGn_r, YlOrBr, YlOrBr_r, YlOrRd, YlOrRd_r, afmhot, afmhot_r, autumn, autumn_r, binary, binary_r, bone, bone_r, brg, brg_r, bwr, bwr_r, cividis, cividis_r, cool, cool_r, coolwarm, coolwarm_r, copper, copper_r, cubehelix, cubehelix_r, flag, flag_r, gist_earth, gist_earth_r, gist_gray, gist_gray_r, gist_heat, gist_heat_r, gist_ncar, gist_ncar_r, gist_rainbow, gist_rainbow_r,
gist_stern, gist_stern_r, gist_yarg, gist_yarg_r, gnuplot, gnuplot2, gnuplot2_r, gnuplot_r, gray, gray_r, hot, hot_r, hsv, hsv_r, icefire, icefire_r, inferno, inferno_r, jet, jet_r, magma, magma_r, mako, mako_r, nipy_spectral, nipy_spectral_r, ocean, ocean_r, pink, pink_r, plasma, plasma_r, prism, prism_r, rainbow, rainbow_r, rocket, rocket_r, seismic, seismic_r, spring, spring_r, summer, summer_r, tab10, tab10_r, tab20, tab20_r, tab20b, tab20b_r, tab20c, tab20c_r, terrain, terrain_r,
twilight, twilight_r, twilight_shifted, twilight_shifted_r, viridis, viridis_r, vlag, vlag_r, winter, winter_r</code></pre></div>
<p>A simple venn diagram.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">from</span> matplotlib_venn <span class="im">import</span> venn2
venn2(subsets <span class="op">=</span> (<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</code></pre></div>
<p>A more complicated venn diagram.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt
<span class="im">import</span> numpy <span class="im">as</span> np
<span class="im">from</span> matplotlib_venn <span class="im">import</span> venn3, venn3_circles
plt.figure(figsize<span class="op">=</span>(<span class="dv">4</span>,<span class="dv">4</span>))
v <span class="op">=</span> venn3(subsets<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>), set_labels <span class="op">=</span> (<span class="st">&#39;A&#39;</span>, <span class="st">&#39;B&#39;</span>, <span class="st">&#39;C&#39;</span>))
v.get_patch_by_id(<span class="st">&#39;100&#39;</span>).set_alpha(<span class="fl">1.0</span>)
v.get_patch_by_id(<span class="st">&#39;100&#39;</span>).set_color(<span class="st">&#39;white&#39;</span>)
v.get_label_by_id(<span class="st">&#39;100&#39;</span>).set_text(<span class="st">&#39;Unknown&#39;</span>)
v.get_label_by_id(<span class="st">&#39;A&#39;</span>).set_text(<span class="st">&#39;Set &quot;A&quot;&#39;</span>)
c <span class="op">=</span> venn3_circles(subsets<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>), linestyle<span class="op">=</span><span class="st">&#39;dotted&#39;</span>)
c[<span class="dv">0</span>].set_lw(<span class="fl">1.0</span>)
c[<span class="dv">0</span>].set_ls(<span class="st">&#39;dotted&#39;</span>)
plt.title(<span class="st">&quot;Sample Venn diagram&quot;</span>)
plt.annotate(<span class="st">&#39;Unknown set&#39;</span>, xy<span class="op">=</span>v.get_label_by_id(<span class="st">&#39;100&#39;</span>).get_position() <span class="op">-</span> np.array([<span class="dv">0</span>, <span class="fl">0.05</span>]), xytext<span class="op">=</span>(<span class="op">-</span><span class="dv">70</span>,<span class="op">-</span><span class="dv">70</span>),
             ha<span class="op">=</span><span class="st">&#39;center&#39;</span>, textcoords<span class="op">=</span><span class="st">&#39;offset points&#39;</span>, bbox<span class="op">=</span><span class="bu">dict</span>(boxstyle<span class="op">=</span><span class="st">&#39;round,pad=0.5&#39;</span>, fc<span class="op">=</span><span class="st">&#39;gray&#39;</span>, alpha<span class="op">=</span><span class="fl">0.1</span>),
                          arrowprops<span class="op">=</span><span class="bu">dict</span>(arrowstyle<span class="op">=</span><span class="st">&#39;-&gt;&#39;</span>, connectionstyle<span class="op">=</span><span class="st">&#39;arc3,rad=0.5&#39;</span>,color<span class="op">=</span><span class="st">&#39;gray&#39;</span>))
                          plt.show()</code></pre></div>
<p>Log scales seem to always be a challenge. Here is at least one solution to change ticks to log manually.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">y_major_ticks <span class="op">=</span> [np.log(<span class="dv">100</span>),np.log(<span class="dv">200</span>),np.log(<span class="dv">300</span>),np.log(<span class="dv">400</span>),np.log(<span class="dv">500</span>),np.log(<span class="dv">600</span>),np.log(<span class="dv">700</span>),np.log(<span class="dv">800</span>),np.log(<span class="dv">900</span>),<span class="op">\</span>
                 np.log(<span class="dv">1000</span>),np.log(<span class="dv">2000</span>),np.log(<span class="dv">3000</span>),np.log(<span class="dv">4000</span>),np.log(<span class="dv">5000</span>),np.log(<span class="dv">6000</span>),np.log(<span class="dv">7000</span>),np.log(<span class="dv">8000</span>),np.log(<span class="dv">9000</span>),<span class="op">\</span>
                 np.log(<span class="dv">10000</span>),np.log(<span class="dv">20000</span>),np.log(<span class="dv">30000</span>),np.log(<span class="dv">40000</span>),np.log(<span class="dv">50000</span>),np.log(<span class="dv">60000</span>),np.log(<span class="dv">70000</span>),np.log(<span class="dv">80000</span>),np.log(<span class="dv">90000</span>),<span class="op">\</span>
                 np.log(<span class="dv">100000</span>),np.log(<span class="dv">200000</span>),np.log(<span class="dv">300000</span>),np.log(<span class="dv">400000</span>),np.log(<span class="dv">500000</span>),np.log(<span class="dv">600000</span>),np.log(<span class="dv">700000</span>),np.log(<span class="dv">800000</span>),np.log(<span class="dv">900000</span>),<span class="op">\</span>
                 np.log(<span class="dv">1000000</span>),np.log(<span class="dv">2000000</span>),np.log(<span class="dv">3000000</span>),np.log(<span class="dv">4000000</span>),np.log(<span class="dv">5000000</span>),np.log(<span class="dv">6000000</span>),np.log(<span class="dv">7000000</span>),np.log(<span class="dv">8000000</span>),np.log(<span class="dv">9000000</span>),<span class="op">\</span>
                 np.log(<span class="dv">10000000</span>)]

y_major_tick_labels <span class="op">=</span> [<span class="st">&quot;100&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>, <span class="st">&quot;1000&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>, <span class="st">&quot;10,000&quot;</span>,<span class="op">\</span>
                       <span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;100,000&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>, <span class="st">&quot;1,000,000&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>, <span class="st">&quot;10,000,000&quot;</span> ]
ax1.set_yticks(y_major_ticks)
ax1.set_yticklabels(y_major_tick_labels, fontsize <span class="op">=</span> axisfont)
ax1.yaxis.set_tick_params(width<span class="op">=</span>scale, color <span class="op">=</span> grey3, length <span class="op">=</span> <span class="dv">6</span>)</code></pre></div>
</div>
<div id="seaborn" class="section level2">
<h2><span class="header-section-number">4.3</span> Seaborn</h2>
<p>Here is a general bar plot that includes some commonly used parameters.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="co"># fits my 22 inch monitor</span>
plt.figure(figsize<span class="op">=</span>(<span class="fl">19.17</span>,<span class="fl">11.98</span>))
<span class="co"># order controls the display order of the samples</span>
sns.catplot(x<span class="op">=</span><span class="st">&quot;Sample&quot;</span>, y<span class="op">=</span><span class="st">&quot;Somatic&quot;</span>, kind<span class="op">=</span><span class="st">&quot;bar&quot;</span>, data<span class="op">=</span>var_counts, order<span class="op">=</span>labels)<span class="op">;</span>
<span class="co"># keeps x-axis labels, but eliminates the tick mark</span>
plt.tick_params(labelbottom<span class="op">=</span><span class="va">True</span>, bottom<span class="op">=</span><span class="va">False</span>)
<span class="co"># trim off the x-axis</span>
sns.despine(offset<span class="op">=</span><span class="dv">10</span>, trim<span class="op">=</span><span class="va">True</span>, bottom<span class="op">=</span><span class="va">True</span>)
<span class="co"># labels</span>
plt.title(<span class="st">&#39;&#39;</span>)
plt.ylabel(<span class="st">&#39;&#39;</span>, fontsize<span class="op">=</span><span class="dv">8</span>)
plt.xlabel(<span class="st">&#39;&#39;</span>, fontsize<span class="op">=</span><span class="dv">8</span>)
<span class="co"># manual control of xlabels</span>
labels <span class="op">=</span> [<span class="st">&#39;Indiv_1-a&#39;</span>,<span class="st">&#39;Indiv_2&#39;</span>,<span class="st">&#39;Indiv_3&#39;</span>,<span class="st">&#39;Indiv_1-b&#39;</span>]
<span class="co"># control xtick order</span>
plt.xticks(<span class="bu">range</span>(<span class="bu">len</span>(labels)), labels, rotation<span class="op">=</span><span class="dv">45</span>)
<span class="co"># control the number of x-ticks</span>
plt.locator_params(axis<span class="op">=</span><span class="st">&#39;x&#39;</span>, nbins<span class="op">=</span><span class="dv">10</span>)
<span class="co"># legend positioning</span>
plt.legend(loc<span class="op">=</span><span class="st">&#39;upper right&#39;</span>)
<span class="co"># log scale</span>
plt.gca().set_yscale(<span class="st">&#39;log&#39;</span>)
<span class="co"># this is better if neg values are needed</span>
plt.gca().set_yscale(<span class="st">&#39;symlog&#39;</span>)
<span class="co"># fit plot to display</span>
plt.tight_layout()
plt.show()
<span class="co"># save figure with tight_layout</span>
plt.savefig(<span class="st">&quot;test.svg&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;svg&quot;</span>, bbox_inches<span class="op">=</span><span class="st">&quot;tight&quot;</span>, dpi<span class="op">=</span><span class="dv">1000</span>)</code></pre></div>
<p>Signifance information can be added by including p-values and label bars using the following code.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">x1, x2 <span class="op">=</span> <span class="dv">0</span>, <span class="dv">1</span> <span class="co"># columns to annotate on the plot</span>
y2, y1 <span class="op">=</span> <span class="dv">20</span>, <span class="dv">15</span> <span class="co"># placement of the line and how for down the vertical legs go</span>
plt.plot([x1,x1, x2, x2], [y1, y2, y2, y1], linewidth<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">&#39;k&#39;</span>) <span class="co"># stats line</span>
plt.text((x1<span class="op">+</span>x2)<span class="op">*</span>.<span class="dv">5</span>, y2<span class="op">+</span><span class="dv">2</span>, <span class="st">&quot;p=0.09&quot;</span>, ha<span class="op">=</span><span class="st">&#39;center&#39;</span>, va<span class="op">=</span><span class="st">&#39;bottom&#39;</span>, fontsize<span class="op">=</span><span class="dv">8</span>) <span class="co"># p-value or sig</span></code></pre></div>
</div>
<div id="statistics" class="section level2">
<h2><span class="header-section-number">4.4</span> Statistics</h2>
<p>This is a two-sided T-test for the null hypothesis that two populations have the same means. It is important to note that it assumes the population variances are the same, so this must be changed if the assumption is incorrect.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="co"># ttest_ind(a, b, axis=0, equal_var=True, nan_policy=&#39;propagate&#39;)</span>
<span class="im">from</span> scipy.stats <span class="im">import</span> ttest_ind
ttest_ind(df[df[<span class="st">&#39;sample&#39;</span>] <span class="op">==</span> <span class="st">&#39;one&#39;</span>][<span class="st">&#39;means&#39;</span>], df[df[<span class="st">&#39;sample&#39;</span>] <span class="op">==</span> <span class="st">&#39;two&#39;</span>][<span class="st">&#39;means&#39;</span>])</code></pre></div>
</div>
<div id="various-plot-styles" class="section level2">
<h2><span class="header-section-number">4.5</span> Various Plot Styles</h2>
<p>This displays each individual datapoint overlayed on a boxplot</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">ax <span class="op">=</span> sns.boxplot(x<span class="op">=</span><span class="st">&#39;day&#39;</span>, y<span class="op">=</span><span class="st">&#39;total_bill&#39;</span>, data<span class="op">=</span>tips)
ax <span class="op">=</span> sns.swarmplot(x<span class="op">=</span><span class="st">&#39;day&#39;</span>, y<span class="op">=</span><span class="st">&#39;total_bill&#39;</span>, data<span class="op">=</span>tips, color<span class="op">=</span><span class="st">&#39;.25&#39;</span>)</code></pre></div>
<!--chapter:end:03-visualization.Rmd-->
</div>
</div>
<div id="biology" class="section level1">
<h1><span class="header-section-number">5</span> Biology</h1>
<div id="general" class="section level2">
<h2><span class="header-section-number">5.1</span> General</h2>
<p>Some helpful commands for genetic sequence.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">from</span> string <span class="im">import</span> ascii_uppercase <span class="co"># python 3</span>
<span class="im">from</span> string <span class="im">import</span> upper, lower <span class="co"># python 2</span>
upper(<span class="st">&#39;tcga&#39;</span>)
lower(<span class="st">&#39;TCGA&#39;</span>)
title(<span class="st">&#39;tcga&#39;</span>) <span class="co"># capitalize the first letter</span></code></pre></div>
</div>
<div id="biopython" class="section level2">
<h2><span class="header-section-number">5.2</span> Biopython</h2>
<p>Reverse complement of sequence</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">from</span> Bio.Seq <span class="im">import</span> Seq
<span class="bu">str</span>(Seq(i).reverse_complement())</code></pre></div>
</div>
<div id="ucsc-genome-browser" class="section level2">
<h2><span class="header-section-number">5.3</span> UCSC Genome Browser</h2>
<p>Get sequence from UCSC genome browser</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">from</span> subprocess <span class="im">import</span> check_output, STDOUT
temp <span class="op">=</span> check_output(<span class="st">&#39;wget -qO- http://genome.ucsc.edu/cgi-bin/das/hg19/dna?segment=</span><span class="sc">%s</span><span class="st">:</span><span class="sc">%s</span><span class="st">,</span><span class="sc">%s</span><span class="st">&#39;</span> <span class="op">%</span> (vcfObj.chrom,low,high), stderr<span class="op">=</span>STDOUT, shell<span class="op">=</span><span class="va">True</span>)</code></pre></div>
</div>
<div id="ref-genome" class="section level2">
<h2><span class="header-section-number">5.4</span> Ref Genome</h2>
<p>Get sequence from reference genome</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">from</span> subprocess <span class="im">import</span> check_output, STDOUT
temp <span class="op">=</span> check_output(<span class="st">&#39;samtools faidx </span><span class="sc">%s</span><span class="st"> </span><span class="sc">%s</span><span class="st">:</span><span class="sc">%s</span><span class="st">-</span><span class="sc">%s</span><span class="st">&#39;</span> <span class="op">%</span> (ref, vcfObj.chrom, low, high), stderr<span class="op">=</span>STDOUT, shell<span class="op">=</span><span class="va">True</span>)

finalSeq <span class="op">=</span> <span class="st">&#39;&#39;</span>
<span class="cf">for</span> line <span class="kw">in</span> temp.decode(<span class="st">&#39;UTF-8&#39;</span>).split(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>):
<span class="cf">for</span> line <span class="kw">in</span> temp.decode(<span class="st">&#39;UTF-8&#39;</span>).split(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>): <span class="co"># this is only necessary in python 3 to convert binary to string</span>
    <span class="cf">if</span> <span class="st">&#39;&gt;&#39;</span> <span class="kw">not</span> <span class="kw">in</span> line:
        finalSeq <span class="op">+=</span> line

finalSeq <span class="op">=</span> finalSeq.upper()</code></pre></div>
</div>
<div id="personal-information" class="section level2">
<h2><span class="header-section-number">5.5</span> Personal Information</h2>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="co"># parse vcf file with parseline</span>
<span class="cf">if</span> <span class="st">&#39;#&#39;</span> <span class="kw">not</span> <span class="kw">in</span> line <span class="kw">and</span> <span class="st">&#39;chr&#39;</span> <span class="kw">in</span> line: <span class="co"># skip the info</span>
<span class="co"># vcf handling</span>
<span class="im">from</span> parseline <span class="im">import</span> VCFObj
<span class="co"># or</span>
<span class="im">from</span> util <span class="im">import</span> VCFObj
vcfObj <span class="op">=</span> VCFObj(vcfLine)
<span class="co"># available attributes: ao, dp, af, wt, var, chrom, location</span></code></pre></div>
<!--chapter:end:04-biology.Rmd-->
</div>
</div>
<div id="io" class="section level1">
<h1><span class="header-section-number">6</span> Data I/O</h1>
<div id="reading-data-files" class="section level2">
<h2><span class="header-section-number">6.1</span> Reading Data Files</h2>
<p>Opening .gz files</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> gzip
<span class="cf">for</span> line <span class="kw">in</span> gzip.<span class="bu">open</span>(<span class="st">&#39;myFile.gz&#39;</span>):
    <span class="bu">print</span> line</code></pre></div>
</div>
<div id="pickles" class="section level2">
<h2><span class="header-section-number">6.2</span> Pickles</h2>
<p>Writing data in pickle format</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> pickle
p <span class="op">=</span> <span class="bu">open</span>(<span class="st">&#39;principle.pkl&#39;</span>, <span class="st">&#39;wb&#39;</span>)
pickle.dump(principleData, p)
p.close()</code></pre></div>
<p>Reading data in pickle format</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> pickle
p <span class="op">=</span> <span class="bu">open</span>(<span class="st">&#39;principle.pkl&#39;</span>, <span class="st">&#39;rb&#39;</span>)
principleData <span class="op">=</span> pickle.load(p)
p.close()</code></pre></div>
<!--chapter:end:05-IO.Rmd-->
</div>
</div>
<div id="pandas" class="section level1">
<h1><span class="header-section-number">7</span> Pandas</h1>
<div id="file-io" class="section level2">
<h2><span class="header-section-number">7.1</span> File I/O</h2>
<p>Read a csv file into a DataFrame.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">pd.read_csv(filepath)</code></pre></div>
<p>Write a DataFrame to a file.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">x.to_csv(path_or_buf<span class="op">=</span><span class="st">&#39;outputDir&#39;</span>, sep<span class="op">=</span><span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, header<span class="op">=</span><span class="va">False</span>, index<span class="op">=</span><span class="va">False</span>)</code></pre></div>
</div>
<div id="relabeling" class="section level2">
<h2><span class="header-section-number">7.2</span> Relabeling</h2>
<p>Rename a column or group of columns can be done by passing a dictionary of the changes.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">    df <span class="op">=</span> df.rename(columns<span class="op">=</span>{<span class="st">&#39;a&#39;</span>:<span class="st">&#39;b&#39;</span>,<span class="st">&#39;c&#39;</span>:<span class="st">&#39;d&#39;</span>})</code></pre></div>
</div>
<div id="sorting-and-arranging" class="section level2">
<h2><span class="header-section-number">7.3</span> Sorting and Arranging</h2>
<p>The data in a DataFrame can be sorted in numeric or lexicographic order. The following code sorts the values within the columns a and b.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">df.sort_values([<span class="st">&#39;a&#39;</span>,<span class="st">&#39;b&#39;</span>], ascending<span class="op">=</span><span class="va">False</span>)</code></pre></div>
</div>
<div id="editing-data" class="section level2">
<h2><span class="header-section-number">7.4</span> Editing Data</h2>
<p>Drop columns from a DataFrame.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> numpy <span class="im">as</span> np
df <span class="op">=</span> pd.DataFrame(np.arange(<span class="dv">12</span>).reshape(<span class="dv">3</span>,<span class="dv">4</span>),
                    columns<span class="op">=</span>[<span class="st">&#39;A&#39;</span>, <span class="st">&#39;B&#39;</span>, <span class="st">&#39;C&#39;</span>, <span class="st">&#39;D&#39;</span>])
<span class="bu">print</span>(df)

df <span class="op">=</span> df.drop(columns<span class="op">=</span>[<span class="st">&#39;B&#39;</span>, <span class="st">&#39;C&#39;</span>]) <span class="co"># may not work in python 2</span>
df <span class="op">=</span> df.drop([<span class="st">&#39;B&#39;</span>, <span class="st">&#39;C&#39;</span>], axis<span class="op">=</span><span class="dv">1</span>) <span class="co"># this works in python 2</span>
<span class="bu">print</span>(df)</code></pre></div>
<p>Changing the datatype of a column of data can be done by just changing column type.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">df.Age <span class="op">=</span> df.Age.astype(<span class="bu">str</span>)</code></pre></div>
<div id="replace-values" class="section level3">
<h3><span class="header-section-number">7.4.1</span> Replace values</h3>
<p>New data can be set within a DataFrame one subset at a time in a way that will avoid the SettingWithCopyWarning.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> pandas <span class="im">as</span> pd
df <span class="op">=</span> pd.DataFrame({<span class="st">&#39;Trait&#39;</span>:[<span class="st">&#39;Seed_Shape&#39;</span>,<span class="st">&#39;Seed_Shape&#39;</span>,<span class="st">&#39;Flower_Color&#39;</span>,<span class="st">&#39;Flower_Color&#39;</span>],
                    <span class="st">&#39;Phenotype&#39;</span>:[<span class="st">&#39;Round&#39;</span>,<span class="st">&#39;Wrinkled&#39;</span>,<span class="st">&#39;Purple&#39;</span>,<span class="st">&#39;White&#39;</span>]})
df.loc[df.Trait <span class="op">==</span> <span class="st">&#39;Seed_Shape&#39;</span>, <span class="st">&#39;Affected_Part&#39;</span>] <span class="op">=</span> <span class="st">&#39;Seed&#39;</span>
df.loc[df.Trait <span class="op">==</span> <span class="st">&#39;Flower_Color&#39;</span>, <span class="st">&#39;Affected_Part&#39;</span>] <span class="op">=</span> <span class="st">&#39;Flower&#39;</span>
<span class="bu">print</span>(df)</code></pre></div>
<pre><code>##           Trait Phenotype Affected_Part
## 0    Seed_Shape     Round          Seed
## 1    Seed_Shape  Wrinkled          Seed
## 2  Flower_Color    Purple        Flower
## 3  Flower_Color     White        Flower</code></pre>
<p>There is a more simple alternative to the above method buit it may result in the SettingWithCopyWarning.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">df <span class="op">=</span> df.replace(<span class="st">&#39;pork&#39;</span>,<span class="st">&#39;bacon&#39;</span>)</code></pre></div>
</div>
</div>
<div id="combining-data-structures" class="section level2">
<h2><span class="header-section-number">7.5</span> Combining Data Structures</h2>
<p>The following merges df and df2 using inner to get the intersection on the Sample column, where indexes are ignored if the merging is performed on a column as in the following example. The other possible merging strategies are: left: use only keys from left frame, similar to a SQL left outer join; preserve key order. right: use only keys from right frame, similar to a SQL right outer join; preserve key order. outer: use union of keys from both frames, similar to a SQL full outer join; sort keys lexicographically. inner: use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">df <span class="op">=</span> pd.merge(df, df2, how<span class="op">=</span><span class="st">&#39;inner&#39;</span>, on<span class="op">=</span>[<span class="st">&#39;Sample&#39;</span>])</code></pre></div>
<p>Appending to a Dataframe attaches a DataFrame after another one.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">df <span class="op">=</span> pd.DataFrame([[<span class="dv">1</span>, <span class="dv">2</span>], [<span class="dv">3</span>, <span class="dv">4</span>]], columns<span class="op">=</span><span class="bu">list</span>(<span class="st">&#39;AB&#39;</span>))
df2 <span class="op">=</span> pd.DataFrame([[<span class="dv">5</span>, <span class="dv">6</span>], [<span class="dv">7</span>, <span class="dv">8</span>]], columns<span class="op">=</span><span class="bu">list</span>(<span class="st">&#39;AB&#39;</span>))
df.append(df2)</code></pre></div>
</div>
<div id="splitting" class="section level2">
<h2><span class="header-section-number">7.6</span> Splitting</h2>
<p>Remove duplicates</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">x <span class="op">=</span> x[<span class="op">~</span>x.index.duplicated(keep<span class="op">=</span><span class="st">&#39;first&#39;</span>)] <span class="co"># most ideal method</span>

data <span class="op">=</span> pd.DataFrame({<span class="st">&#39;k1&#39;</span>:[<span class="st">&#39;one&#39;</span>,<span class="st">&#39;two&#39;</span>]<span class="op">*</span><span class="dv">3</span><span class="op">+</span>[<span class="st">&#39;two&#39;</span>],<span class="st">&#39;k2&#39;</span>:[<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">4</span>]})
data.duplicated() <span class="co"># identify duplicate data</span>
data[â€˜k1â€™].duplicated()
data[<span class="st">&#39;k1&#39;</span>].drop_duplicates()
data.drop_duplicates[<span class="st">&#39;k1&#39;</span>] <span class="co"># this does the same thing as the previous line</span>
data.drop_duplicates([<span class="st">&#39;k1&#39;</span>,<span class="st">&#39;k2&#39;</span>], keep<span class="op">=</span><span class="st">&#39;last&#39;</span>) <span class="co"># drops unique found in k1 and k2 and keeps the last indexed duplicate</span></code></pre></div>
</div>
<div id="summarizing" class="section level2">
<h2><span class="header-section-number">7.7</span> Summarizing</h2>
<p>The mean of column values can be calculated where each of the columns is grouped by the data in a specified column.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">temp[[<span class="st">&#39;Sample&#39;</span>,<span class="st">&#39;VAF&#39;</span>,<span class="st">&#39;Var_Count&#39;</span>]].groupby(<span class="st">&#39;Sample&#39;</span>).mean()</code></pre></div>
</div>
<div id="arithmetic-and-row-wise-analysis" class="section level2">
<h2><span class="header-section-number">7.8</span> Arithmetic and Row-wise Analysis</h2>
<p>Sometimes it is helpful to analyze the value in a particular cell in a conditional manner depending on itâ€™s value and then set the result of this analysis to a corresponding cell in a new column. Here is an example where the VAF of a variant is conditionally analyzed</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="kw">def</span> LOH(x):
    <span class="cf">if</span> x <span class="op">&gt;</span> <span class="fl">0.75</span>: <span class="cf">return</span> <span class="dv">1</span> <span class="op">-</span> x
    <span class="cf">elif</span> x <span class="op">&lt;=</span> <span class="fl">0.75</span> <span class="kw">and</span> x <span class="op">&gt;</span> <span class="fl">0.25</span>: <span class="cf">return</span> <span class="bu">abs</span>(<span class="fl">0.5</span> <span class="op">-</span> x)
    <span class="cf">else</span>: <span class="cf">return</span> <span class="dv">0</span>
all_vars[<span class="st">&#39;LOH&#39;</span>] <span class="op">=</span> all_vars.VAF.transform(LOH)
max_loh <span class="op">=</span> all_vars.groupby(<span class="st">&#39;Sample&#39;</span>).LOH.<span class="bu">max</span>().reset_index().rename(columns<span class="op">=</span>{<span class="st">&#39;LOH&#39;</span>:<span class="st">&#39;Max_LOH&#39;</span>})
all_vars <span class="op">=</span> pd.merge(all_vars, max_loh, how<span class="op">=</span><span class="st">&#39;inner&#39;</span>, on<span class="op">=</span>[<span class="st">&#39;Sample&#39;</span>])</code></pre></div>
<!--chapter:end:06-pandas.Rmd-->
</div>
</div>
<div id="git" class="section level1">
<h1><span class="header-section-number">8</span> Git</h1>
<div id="setup" class="section level2">
<h2><span class="header-section-number">8.1</span> Setup</h2>
<div id="git-setup" class="section level3">
<h3><span class="header-section-number">8.1.1</span> Git Setup</h3>
<p>The username and email needs to be added after git is installed.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> config --global user.name <span class="st">&quot;me&quot;</span>
<span class="fu">git</span> config --global user.email <span class="st">&quot;me@gmail.com&quot;</span></code></pre></div>
<p>After this information has been set, it can be checked.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> config --list</code></pre></div>
</div>
<div id="repository-initiation" class="section level3">
<h3><span class="header-section-number">8.1.2</span> Repository Initiation</h3>
<p>To setup a repository, create a folder with an initial file like a README and then initiate it.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> init
<span class="fu">git</span> status</code></pre></div>
</div>
<div id="mirror-on-online-repository" class="section level3">
<h3><span class="header-section-number">8.1.3</span> Mirror on Online Repository</h3>
<p>Create a repository on a repository like github, gitlab, bitbucket, or sourceforge. Then the local git repository can be synched with the online repository.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> remote add origin url-of-online-repository-here
<span class="fu">git</span> push -u origin master</code></pre></div>
<p>Of course the repository could just be setup first and then cloned.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> clone url-of-online-repository-here</code></pre></div>
</div>
</div>
<div id="manipulating-commits" class="section level2">
<h2><span class="header-section-number">8.2</span> Manipulating Commits</h2>
<div id="repository-status" class="section level3">
<h3><span class="header-section-number">8.2.1</span> Repository Status</h3>
<p>The commit history of a repository can be displayed in verbose form and in summarized form.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> log
<span class="fu">git</span> log --oneline</code></pre></div>
</div>
<div id="file-checkout" class="section level3">
<h3><span class="header-section-number">8.2.2</span> File Checkout</h3>
<p>To restore a previous version of a file it can be checked out by first identifying the version to be used using the log history and then restoring the desired file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> log --oneline
<span class="fu">git</span> checkout <span class="op">&lt;</span>commit number<span class="op">&gt;</span> file.txt</code></pre></div>
</div>
<div id="resetting-a-repository" class="section level3">
<h3><span class="header-section-number">8.2.3</span> Resetting a Repository</h3>
<p>To discard the effect of the previous operation on a file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> reset HEAD file.txt</code></pre></div>
<p>The previous version of the a file can then be restored.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> checkout -- file.txt</code></pre></div>
<!--chapter:end:07-git.Rmd-->
</div>
</div>
</div>
<div id="git" class="section level1">
<h1><span class="header-section-number">9</span> Git</h1>
<div id="formatting" class="section level2">
<h2><span class="header-section-number">9.1</span> Formatting</h2>
<p>Automatic newlines are inserted by default; this behavior can be overidden with the following.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">:<span class="ex">set</span> wrap
:<span class="ex">set</span> textwidth=0 wrapmargin=0</code></pre></div>
</div>
<div id="spellcheck" class="section level2">
<h2><span class="header-section-number">9.2</span> Spellcheck</h2>
<p>To setup spellchecking first setup a personal dictionary file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> -p ~/.vim/spell/    # makes a directory for personal dictionary</code></pre></div>
<p>Then refer to the dictionary file within VIM, and enable spellchecking.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># set personal dictionary</span>
:<span class="ex">set</span> spellfile=~/.vim/spell/en.utf-8.add
<span class="co"># turn spellcheck on</span>
:<span class="ex">set</span> spell</code></pre></div>
<p>Get spellcheck commands.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">:<span class="ex">help</span> spell</code></pre></div>
<p>Add a word to personal dictionary.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">zg</span></code></pre></div>
<p>Move to next and previous misspelled word.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">]<span class="ex">s</span>
[<span class="ex">s</span></code></pre></div>
<p>Get suggestions for misspelled word.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="va">z=</span></code></pre></div>
<div id="git-setup-1" class="section level3">
<h3><span class="header-section-number">9.2.1</span> Git Setup</h3>
<p>The username and email needs to be added after git is installed.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> config --global user.name <span class="st">&quot;me&quot;</span>
<span class="fu">git</span> config --global user.email <span class="st">&quot;me@gmail.com&quot;</span></code></pre></div>
<p>After this information has been set, it can be checked.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> config --list</code></pre></div>
</div>
<div id="repository-initiation-1" class="section level3">
<h3><span class="header-section-number">9.2.2</span> Repository Initiation</h3>
<p>To setup a repository, create a folder with an initial file like a README and then initiate it.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> init
<span class="fu">git</span> status</code></pre></div>
</div>
<div id="mirror-on-online-repository-1" class="section level3">
<h3><span class="header-section-number">9.2.3</span> Mirror on Online Repository</h3>
<p>Create a repository on a repository like github, gitlab, bitbucket, or sourceforge. Then the local git repository can be synched with the online repository.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> remote add origin url-of-online-repository-here
<span class="fu">git</span> push -u origin master</code></pre></div>
<p>Of course the repository could just be setup first and then cloned.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> clone url-of-online-repository-here</code></pre></div>
</div>
</div>
<div id="manipulating-commits-1" class="section level2">
<h2><span class="header-section-number">9.3</span> Manipulating Commits</h2>
<div id="repository-status-1" class="section level3">
<h3><span class="header-section-number">9.3.1</span> Repository Status</h3>
<p>The commit history of a repository can be displayed in verbose form and in summarized form.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> log
<span class="fu">git</span> log --oneline</code></pre></div>
</div>
<div id="file-checkout-1" class="section level3">
<h3><span class="header-section-number">9.3.2</span> File Checkout</h3>
<p>To restore a previous version of a file it can be checked out by first identifying the version to be used using the log history and then restoring the desired file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> log --oneline
<span class="fu">git</span> checkout <span class="op">&lt;</span>commit number<span class="op">&gt;</span> file.txt</code></pre></div>
</div>
<div id="resetting-a-repository-1" class="section level3">
<h3><span class="header-section-number">9.3.3</span> Resetting a Repository</h3>
<p>To discard the effect of the previous operation on a file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> reset HEAD file.txt</code></pre></div>
<p>The previous version of the a file can then be restored.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">git</span> checkout -- file.txt</code></pre></div>
<!--chapter:end:08-VIM.Rmd-->
</div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<!--chapter:end:99-references.Rmd-->
<div id="refs" class="references">
<div id="ref-R-bookdown">
<p>Xie, Yihui. 2018. <em>Bookdown: Authoring Books and Technical Documents with R Markdown</em>. <a href="https://CRAN.R-project.org/package=bookdown" class="uri">https://CRAN.R-project.org/package=bookdown</a>.</p>
</div>
</div>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
